tokio/src/time/interval.rs

```diff
@@ -521,6 +521,103 @@ impl Interval {
         self.delay.as_mut().reset(Instant::now() + self.period);
     }
 
+    /// Resets the interval immediately.
+    ///
+    /// This method ignores [`MissedTickBehavior`] strategy.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::time;
+    ///
+    /// use std::time::Duration;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let mut interval = time::interval(Duration::from_millis(100));
+    ///
+    ///     interval.tick().await;
+    ///
+    ///     time::sleep(Duration::from_millis(50)).await;
+    ///     interval.reset_immediately();
+    ///
+    ///     interval.tick().await;
+    ///     interval.tick().await;
+    ///
+    ///     // approximately 150ms have elapsed.
+    /// }
+    /// ```
+    pub fn reset_immediately(&mut self) {
+        self.delay.as_mut().reset(Instant::now());
+    }
+
+    /// Resets the interval after the specified [`std::time::Duration`].
+    ///
+    /// This method ignores [`MissedTickBehavior`] strategy.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::time;
+    ///
+    /// use std::time::Duration;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let mut interval = time::interval(Duration::from_millis(100));
+    ///     interval.tick().await;
+    ///
+    ///     time::sleep(Duration::from_millis(50)).await;
+    ///
+    ///     let after = Duration::from_millis(20);
+    ///     interval.reset_after(after);
+    ///
+    ///     interval.tick().await;
+    ///     interval.tick().await;
+    ///
+    ///     // approximately 170ms have elapsed.
+    /// }
+    /// ```
+    pub fn reset_after(&mut self, after: Duration) {
+        self.delay.as_mut().reset(Instant::now() + after);
+    }
+
+    /// Resets the interval to a [`crate::time::instant::Instant`] deadline.
+    /// If deadline is in the past it behaves like `reset_immediately`.
+    ///
+    /// This method ignores [`MissedTickBehavior`] strategy.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::time::{self, Instant};
+    ///
+    /// use std::time::Duration;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let mut interval = time::interval(Duration::from_millis(100));
+    ///     interval.tick().await;
+    ///
+    ///     time::sleep(Duration::from_millis(50)).await;
+    ///
+    ///     let deadline = Instant::now() + Duration::from_millis(30);
+    ///     interval.reset_at(deadline);
+    ///
+    ///     interval.tick().await;
+    ///     interval.tick().await;
+    ///
+    ///     // approximately 180ms have elapsed.
+    /// }
+    /// ```
+    pub fn reset_at(&mut self, deadline: Instant) {
+        if deadline > Instant::now() {
+            self.delay.as_mut().reset(deadline);
+        } else {
+            self.reset_immediately();
+        }
+    }
```

Comment:
What happens if we just do `self.delay.as_mut().reset(deadline)` always? I guess we trigger the missed tick behavior?

Reply:
Using the example code:
```rust
use tokio::time::{self, Instant};

use std::time::Duration;

#[tokio::main]
async fn main() {
    let start_time = Instant::now();
    let mut interval = time::interval(Duration::from_millis(100));

    interval.tick().await;
    println!("Expected: `0s` elapsed, actual: `{:?}`", start_time.elapsed());

    time::sleep(Duration::from_millis(50)).await;
    println!("Expected: `50s` elapsed, actual: `{:?}`", start_time.elapsed());

    interval.reset_at(Instant::now() - Duration::from_millis(220));

    interval.tick().await;
    println!("Expected: `50s` elapsed, actual: `{:?}`", start_time.elapsed());

    interval.tick().await;
    println!("Expected: `100s + 50s = 150s` elapsed, actual: `{:?}`", start_time.elapsed());

    interval.tick().await;
    println!("Expected: `100s + 150s = 250s` elapsed, actual: `{:?}`", start_time.elapsed());

    interval.tick().await;
    println!("Expected: `100s + 250s = 350s` elapsed, actual: `{:?}`", start_time.elapsed());
}
```

Currently we get:
```sh
Expected: `0s` elapsed, actual: `1.209522ms`
Expected: `50s` elapsed, actual: `52.458844ms`
Expected: `50s` elapsed, actual: `53.605279ms`
Expected: `100s + 50s = 150s` elapsed, actual: `153.223861ms`
Expected: `100s + 150s = 250s` elapsed, actual: `253.889257ms`
Expected: `100s + 250s = 350s` elapsed, actual: `353.632062ms`
```

Using `self.delay.as_mut().reset(deadline)` always we get:
```sh
Expected: `0s` elapsed, actual: `1.193677ms`
Expected: `50s` elapsed, actual: `52.362525ms`
Expected: `50s` elapsed, actual: `52.391208ms`
Expected: `100s + 50s = 150s` elapsed, actual: `52.401491ms`
Expected: `100s + 150s = 250s` elapsed, actual: `52.411412ms`
Expected: `100s + 250s = 350s` elapsed, actual: `134.023993ms`
```

If the deadline is in the past, `time_diff.div_euclid(period)` number of ticks will occur simultaneously, and the next tick will occur `time_diff.rem_euclid(period)` after we called `reset_at`.

I think the behavior where the `reset_at` behaves likes `reset_immediately` would be better if `deadline` is in the past.